overlord/devicestate/devicestate_test.go

```diff
@@ -2338,3 +2490,66 @@ func (s *deviceMgrSuite) TestMarkSeededInConfig(c *C) {
 	// changes
 	c.Check(s.state.Changes(), HasLen, 1)
 }
+
+func (s *deviceMgrSuite) TestNewEnoughProxy(c *C) {
+	expectedUserAgent := httputil.UserAgent()
+	log, restore := logger.MockLogger()
+	defer restore()
+	os.Setenv("SNAPD_DEBUG", "1")
+	defer os.Unsetenv("SNAPD_DEBUG")
+
+	expecteds := []string{
+		`parse .*`, // this one won't actually reach the server
+		`Head http://\S+: EOF`,
+		`Head request returned 403 Forbidden.`,
+		`Bogus Snap-Store-Version header "5pre1".`,
+		``,
+	}
+
+	n := 0
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c.Check(r.Header.Get("User-Agent"), Equals, expectedUserAgent)
+		n++
+		switch n {
+		case 1:
+			conn, _, err := w.(http.Hijacker).Hijack()
+			c.Assert(err, IsNil)
+			conn.Close()
+		case 2:
+			w.WriteHeader(403)
+		case 3:
+			w.Header().Set("Snap-Store-Version", "5pre1")
+			w.WriteHeader(200)
+		case 4:
+			w.Header().Set("Snap-Store-Version", "5")
+			w.WriteHeader(200)
+		case 5:
+			w.Header().Set("Snap-Store-Version", "6")
+			w.WriteHeader(200)
+		default:
+			c.Errorf("expected %d results, now on %d", len(expecteds), n)
+		}
+	}))
+	defer server.Close()
+
+	u1 := &url.URL{Opaque: "%a"} // bad urls are bad
```

Comment:
Ok, I get it now. I wonder if it might be easier to read if the "parser error" becomes a separate test? It will also be pretty short, something like:
```
func (s *deviceMgrSuite) TestNewEnoughProxyParse(c *C) {
	log, restore := logger.MockLogger()
	defer restore()
	os.Setenv("SNAPD_DEBUG", "1")
	defer os.Unsetenv("SNAPD_DEBUG")

	badURL := &url.URL{Opaque: "%a"}
	c.Check(devicestate.NewEnoughProxy(badURL, http.DefaultClient), Equals, false)
	c.Check(log.String(), Matches, "(?m).* DEBUG: Cannot check whether proxy store supports a custom serial vault: parse .*")
}
```
and it means we can remove the "magic" around u1, u2 in the test below.